Add link to agent a stage is running on to stage summary

Hammie opened the pull request:
Adds a line to the console view stage summary, showing which agent the stage is running on. 

Fixes #426 

### Testing done

Ran the plugin on a Jenkins instance with multi-stage pipelines, and so far everything seems to work.

![jenkins_pipeline_agent](https://github.com/user-attachments/assets/5c1d747c-8845-40a1-ad16-0a70b9762d1f)

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

timja commented:
> What does this render if the pipeline has `agent none` and there is no `agent` for the stage either? We have some pipelines like that, where an agentless stage sets job properties conditionally before the job starts waiting for an available agent.

Good find, there's a bug it always displays the text when it should only do it if the agent is running on a stage:

<img width="183" alt="image" src="https://github.com/user-attachments/assets/a55f30d7-8979-4595-bb96-bad6fd522132">

Hammie commented:
I've also been testing it on our production instance, and there are some more edge cases where it is not quite working as expected.

- A parallel stage copies the agent property from its first child stage
- A simple single-stage pipeline also shows no agent (even when already running)

I'll try and fix these and also write some tests for them so we can have some confidence in it working well.

But I'm not an experienced Java dev, so we'll see how that goes...

Hammie commented:
I've added some tests and fixed the known issues.

Not really happy with the actual agent lookup code, but I don't think I can really do it better.

At the very least the tests should make it easier if someone wants to try and rewrite it.

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:
```diff
@@ -0,0 +1,15 @@
+
+const StageNodeLink = ({agent}: StageNodeLinkProps) => {
+    const agentName = agent == "built-in" ? "Jenkins" : agent;
+    const href = `../../../../computer/${agent == "built-in" ? "(built-in)" : agent }/`;
```
I don't think this will be reliable, you could be nested in multi-level folders. you should do it from the root, you can access the relative URL that Jenkins is running on from with (in plain JS, react will be a little different): `document.head.dataset.rooturl`

roughly:
`${document.head.dataset.rooturl}/manage/computer/...`

but you may need a react hook to get the dataset

Hammie replied:
I've updated the code to use the root url. Could not really find a reason why a hook would be needed, so I hope it's good.

timja reviewed (commented):
Code looks pretty good from a read, one improvement needed and I'll do a re-test when I get a chance

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:
```diff
@@ -0,0 +1,22 @@
+}
+
+const StageNodeLink = ({agent}: StageNodeLinkProps) => {
+    const agentName = agent == "built-in" ? "Jenkins" : agent;
```
Doesn't seem to be working the agent name is coming through as `jenkins` and not `Jenkins`:

![image](https://github.com/user-attachments/assets/04b90cfd-afa4-4128-80a4-f1008a69c2bf)

timja commented on src/main/frontend/pipeline-console-view/pipeline-console/main/StageNodeLink.tsx:
```diff
@@ -0,0 +1,22 @@
+function getAgentUrl(name: string) {
+    // Wrap built-in in brackets
+    const id = name == "built-in" ? "(built-in)" : name;
+    const rootPath = document.head.dataset.rootUrl
```
When I run this locally it needs to be all in lowercase:
```suggestion
    const rootPath = document.head.dataset.rooturl
```

timja reviewed (changes requested):
The link I am currently getting when I click to view the agent is:
http://localhost:8085/jenkins/job/junit-attachments-test/job/master/31/pipeline-console/undefined/computer/(built-in)/

(with undefined in the url) and the link doesn't work

timja reviewed (approved):
LGTM thanks for your work on this!
